cis194/week4/tien/homework4/src/Homework4.hs

```diff
@@ -30,3 +30,42 @@ data Tree a
   = Leaf
   | Node Integer (Tree a) a (Tree a)
   deriving (Show, Eq)
+
+checkDepth :: Eq a => Tree a -> Integer
+checkDepth Leaf = (0 :: Integer)
+checkDepth (Node _ Leaf _ Leaf) = 1
+checkDepth (Node _ Leaf _ right@(Node _ _ _ _)) = 1 + checkDepth right
+checkDepth (Node _ left@(Node _ _ _ _) _ Leaf) = 1 + checkDepth left
+checkDepth (Node _ left@(Node _ _ _ _) _ right@(Node _ _ _ _)) =
+  1 + checkDepth left + checkDepth right
+
+insert :: Eq t => t -> Tree t -> Tree t
+insert x Leaf = Node 0 Leaf x Leaf
+insert x (Node i left c Leaf) = Node i left c (Node (i + 1) Leaf x Leaf)
+insert x (Node i Leaf c right) = Node i (Node (i + 1) Leaf x Leaf) c right
+insert x (Node i left c right)
+  | (checkDepth left) > (checkDepth right) = Node i left c (insert x right)
```

Comment:
This works. But here is something to think about. Is it better to still use the heights stored on the Node but treat left height equals to right height as a dedicated case?